File: skills/novel-writer/scripts/status.py
#!/usr/bin/env python3
import argparse
import json
from pathlib import Path

from _novel_lib import ensure_project_dirs, load_json, memory_path, read_recent_timeline


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--project", required=True)
    ap.add_argument("--recent", type=int, default=5)
    args = ap.parse_args()

    project_dir = ensure_project_dirs(args.project)
    state_path = project_dir / "state.json"
    state = load_json(state_path, {})
    timeline = read_recent_timeline(memory_path(args.project, "timeline"), args.recent)

    output = {
        "status": "ok",
        "project": args.project,
        "last_chapter": state.get("last_chapter"),
        "last_title": state.get("last_title"),
        "updated_at": state.get("updated_at"),
        "recent_timeline": timeline,
    }
    print(json.dumps(output, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: skills/novel-writer/ts/status.ts
#!/usr/bin/env npx tsx
/**
 * status.ts
 * Show current project progress: last chapter, recent timeline.
 * Replaces scripts/status.py.
 *
 * Usage:
 *   npx tsx skills/novel-writer/ts/status.ts --project mynovel --recent 5
 */

import fsSync from "node:fs";
import path from "node:path";
import { parseArgs } from "node:util";
import { PROJECTS_DIR } from "./apply-patch.js";

function loadJson(filePath: string, fallback: unknown): any {
  if (!fsSync.existsSync(filePath)) return fallback;
  try {
    return JSON.parse(fsSync.readFileSync(filePath, "utf-8"));
  } catch {
    return fallback;
  }
}

function readRecentTimeline(filePath: string, limit: number): unknown[] {
  if (!fsSync.existsSync(filePath)) return [];
  try {
    const lines = fsSync
      .readFileSync(filePath, "utf-8")
      .split("\n")
      .filter((l) => l.trim());
    return lines.slice(-limit).map((l) => JSON.parse(l));
  } catch {
    return [];
  }
}

function main() {
  const { values } = parseArgs({
    options: {
      project: { type: "string" },
      recent: { type: "string", default: "5" },
    },
    strict: true,
  });

  if (!values.project) {
    console.error("--project is required");
    process.exit(1);
  }

  const project = values.project;
  const recent = parseInt(values.recent!, 10);

  const statePath = path.join(PROJECTS_DIR, project, "state.json");
  const timelinePath = path.join(PROJECTS_DIR, project, "memory", "timeline.jsonl");

  const state = loadJson(statePath, {});
  const timeline = readRecentTimeline(timelinePath, recent);

  const output = {
    status: "ok",
    project,
    last_chapter: state.last_chapter ?? null,
    last_title: state.last_title ?? null,
    updated_at: state.updated_at ?? null,
    chapters_written: state.chapters_written ?? [],
    recent_timeline: timeline,
  };
  console.log(JSON.stringify(output, null, 2));
}

main();
